src/_pytest/main.py

```diff
@@ -558,7 +560,17 @@ def _collectfile(self, path):
                 return ()
         return ihook.pytest_collect_file(path=path, parent=self)
 
+    def _check_visited(self, path):
+        st = path.stat()
```

Comment:
@boxed will not like that.. ;)

You might be interested in https://github.com/pytest-dev/pytest/pull/4237 and the discussion at https://github.com/pytest-dev/pytest/issues/2206.

There is also `_recurse` in the python plugin.

I've also added `seen_dirs` in #4237, and wonder if a combination of `realpath` and this would be better maybe?

Reply:
If you think `stat()` (a single syscall) is costly, why would `realpath()` be any better? (Its implementation is, roughly, split the path on the path separator, and call `os.lstat()` on each component.)

Reply:
Rather annoyingly, `os.DirEntry` (yielded by `os.scandir()`) includes the inode number but not the device number.

Reply:
It was a joke I think. I have been trying to cut down on stat() calls because pytest has made millions of them in some rather simple test scenarios. A single stat() call here and there won't be a big deal. The problem is that there has been very many "just a single" things in pytest from 3.4 to 3.9 and some of those weren't really "single" because they were used in a loop (in a loop!).

So I don't know about this case, but you could try running my test script https://github.com/pytest-dev/pytest/issues/2206#issuecomment-433006348 and see if performance is impacted significantly or not.

Reply:
Yeah, I was joking.
But I also think that since we're doing `realpath` already for symlink-resolving it might not be necessary to do any `stat` anymore on top.

I think this should be rebased on features, and then maybe only small changes are required to make the test pass.